plugin/core/types.py

```diff
@@ -376,10 +376,7 @@ def normalize_text_sync(textsync: Union[None, int, Dict[str, Any]]) -> Dict[str,
     result = {}  # type: Dict[str, Any]
     if isinstance(textsync, int):
         change = {"syncKind": textsync}  # type: Optional[Dict[str, Any]]
-        if textsync > TextDocumentSyncKindNone:
-            result["textDocumentSync"] = {"didOpen": {}, "didClose": {}, "change": change}
-        else:
-            result["textDocumentSync"] = {"change": change}
+        result["textDocumentSync"] = {"didOpen": {}, "save": {}, "didClose": {}, "change": change}
```

Comment:
Because I am still learning the LSP repo, 
I just did a quick search though the LSP spec to see more about `textDocumentSync`
and the closes thing I could find regarding `textDocumentSync` is this interface:
```
export interface TextDocumentSyncOptions {
	/**
	 * Open and close notifications are sent to the server. If omitted open close notification should not
	 * be sent.
	 */
	openClose?: boolean;
	/**
	 * Change notifications are sent to the server. See TextDocumentSyncKind.None, TextDocumentSyncKind.Full
	 * and TextDocumentSyncKind.Incremental. If omitted it defaults to TextDocumentSyncKind.None.
	 */
	change?: number;
	/**
	 * If present will save notifications are sent to the server. If omitted the notification should not be
	 * sent.
	 */
	willSave?: boolean;
	/**
	 * If present will save wait until requests are sent to the server. If omitted the request should not be
	 * sent.
	 */
	willSaveWaitUntil?: boolean;
	/**
	 * If present save notifications are sent to the server. If omitted the notification should not be
	 * sent.
	 */
	save?: boolean | SaveOptions;
}
```

but this object that you modified:
 
`{"didOpen": {}, "save": {}, "didClose": {}, "change": change}`

is not adhering to that interface.
So I am kind of confused why that object has keys: `didOpen`, `save`, `didClose`, `change`,
and not `openClose`, `willSave`, `willSaveWaitUntil`, `change`, `save`.

I should probalby take a closer look at this part of the codebase to figure it out, but at first sight I find it a bit strange.

Reply:
`didOpen` and `didClose` each have their own key/dict because of dynamic registration. A server can register `textDocument/didOpen` and not register `textDocument/didClose`.